Add stacktrace entries to BIF calls from emulator

josevalim opened the pull request:
The goal of this change is to improve debugging of emulator calls. For example, the following code

    rem(1, y)

will error with atom `badarith` when y is 0 and the stacktrace has no entry for `erlang:rem/2`, making such cases very hard to debug. This patch makes it so the stacktrace includes `erlang:rem(1, 0)`.

The following emulator BIFs have been changed so far:

  * band/2
  * bor/2
  * bsl/2
  * bsr/2
  * bxor/2
  * div/2
  * int_div/2
  * rem/2
  * sminus/2
  * splus/2
  * stimes/2

I would like to do this change to other emulator operations, such as tuple operations and system limit errors that happen in the functions above, but before I would like to verify that this patch is sound.

I would also appreciate some feedback on the code. Currently I had to sprinkle `if (Arg(0) != 0) goto jump_f;` because we can only change the register after we are sure we won't jump and I am unsure on the best mechanism to avoid the repeating the same code over and over again.

psyeugenic commented:
Just to lift this a bit higher, should operators be considered functions? On some level they are functions already I guess. In general I like better debugging possibilities so I'll give it a +1 for that!

josevalim commented:
> Just to lift this a bit higher, should operators be considered functions? 

I thought it was already the case since both `bar(A, B) -> A + B.` and `bar(A, B) -> erlang:'+'(A, B).` emit this bytecode in `erts_debug:df/2`:

```
000000001CB7E7F0: i_func_info_IaaI 0 foo bar 2 
000000001CB7E818: i_plus_jIxxd j(0000000000000000) 2 x(0) x(1) x(0) 
000000001CB7E848: return 
```

And if you have any feedback on the C code, please let me know.  The only time I write C is when I am contributing to the VM, so it is always rusty. 😅

psyeugenic commented:
As you can see in your example there is no call to anything. It's an instruction. Compare it to `bar(A,B) -> math:pow(A,B).`

    0000000016665F20: i_func_info_IaaI 0 foo bar 2
    0000000016665F48: allocate_tt 0 2
    0000000016665F58: call_bif_e math:pow/2
    0000000016665F68: deallocate_return_Q 0

Here we clearly have a call.

Another thing to consider is that instructions are untraceable. Erlang has a couple of them, for instance `erlang:get/1` is untraceable. Though very much a function it is converted to an instruction in the loader. In my mind functions converted to instructions ceases to be functions at that moment.

As for the C code, you could probably group your similar code into a macro or an inline function. I would wait until the VM team has chance to weigh in though. :)

bjorng commented on erts/emulator/beam/beam_emu.c:
```diff
@@ -1499,7 +1501,11 @@ void process_main(Eterm * x_reg_array, FloatDef* f_reg_array)
      }						\
-     goto lb_Cl_error;				\
+     if (Arg(0) != 0) goto jump_f;		\
+     reg[0] = Op1;				\
```
Not safe if a GC has been done in the arithmetic operation (if the error reason is `system_limit`, there could have been a GC).

Should be:

    reg[0] = reg[live]
    reg[1] = reg[live+1]

bjorng commented on erts/emulator/beam/beam_emu.c:
```diff
@@ -1287,6 +1287,8 @@ void process_main(Eterm * x_reg_array, FloatDef* f_reg_array)
      */
     int neg_o_reds = 0;
 
+    ErtsCodeMFA *mfa_bif = NULL;
```
Introducing a new variable at the top level could be surprisingly expensive, especially for Clang. If at all possible, try to limit the scope of temporary variables by putting them in a block.

bjorng commented on erts/emulator/beam/beam_emu.c:
```diff
@@ -3516,7 +3535,8 @@ do {						\
  find_func_info: {
      SWAPOUT;
-     I = handle_error(c_p, I, reg, NULL);
+     I = handle_error(c_p, I, reg, mfa_bif);
```
This change could be expensive. I would recommend copying the code, and inventing a new label (e.g. `arith_handle_error`).

josevalim commented:
> Another thing to consider is that instructions are untraceable. Erlang has a couple of them, for instance erlang:get/1 is untraceable. Though very much a function it is converted to an instruction in the loader. In my mind functions converted to instructions ceases to be functions at that moment.

What do you mean by untraceable in this case? I have seen some comments around in the source code but I couldn't quite figure out what it means. Does it simply mean no stacktrace entry?

And wouldn't the fact they become instructions be an implementation detail? I was mostly trying to show that operators and functions compile to the same thing (regardless of what it is).

> There should be tests that a garbage collection followed by system_limit works. For inspiration, I suggest looking at big_SUITE.erl and exception_SUITE.erl.

Just to clarify, something along those lines?

```erlang
    Maxbig = maxbig(),
    erlang:garbage_collect(),
    {'EXIT',{system_limit,_}} = (catch Maxbig+1)
```

I am also planning to add a test to exception_SUITE for each of those changes, such that:

```erlang
    {'EXIT', {badarith, [{erlang, '/', [A, B], _} | _]}} = (catch A / B),
    ...
```

Some of the instructions already add the proper bif entry to the stacktrace, such as `apply/3`, so I will make sure they are all listed.

Thanks for all of the feedback, I believe I have enough to continue moving forward. Also, agreed about OTP 20.

garazdawi commented:
>> Another thing to consider is that instructions are untraceable. Erlang has a couple of them, for instance erlang:get/1 is untraceable. Though very much a function it is converted to an instruction in the loader. In my mind functions converted to instructions ceases to be functions at that moment.
>
> What do you mean by untraceable in this case? I have seen some comments around in the source code but I couldn't quite figure out what it means. Does it simply mean no stacktrace entry?

@josevalim untraceable means that it is not possible to use `erlang:trace_pattern/3` to setup a trace on them.

josevalim commented:
I have pushed a new commit that addresses all of @bjorng feedback.

josevalim commented:
Btw, I was going to add the same checks for `apply/2` but it seems there is no BIF_apply_2. At the same time, `erlang:apply/2` seems to be defined in terms of itself:

```erlang
-spec apply(Fun, Args) -> term() when
      Fun :: function(),
      Args :: [term()].
apply(Fun, Args) ->
    erlang:apply(Fun, Args).
```

I suppose it is fully handled by the compiler and we cannot add it to the stacktrace then?

josevalim commented on erts/emulator/beam/beam_emu.c:
```diff
@@ -1798,7 +1815,8 @@ void process_main(Eterm * x_reg_array, FloatDef* f_reg_array)
 	}
-     goto badarg;
+	c_p->freason = BADARG;
+	BIF_ERROR_ARITY_2(make_small(Arg(2)), fast_element_tuple, BIF_element_2);
```
Please review this particular entry. I am not a 100% sure making it small is the way to go.

josevalim commented:
@garazdawi thanks. I have removed the stray comment.

About HiPE, I've added `-compile(native)` to exception_SUITE about 1 hour ago and the test still didn't finish compiling. It is unclear if it is something on my machine or an issue with HiPE.

garazdawi commented:
hmm, that's not very good. Did you try to do it without your changes to the test code?

josevalim commented:
@garazdawi Yes. Here is some more information:

  1. Without my changes, 5 of the 10 tests in exception_SUITE fail when HiPE is enabled. Those tests are related to stacktraces (entries, line numbers, etc), so it seems stacktraces already differ when using HiPE

  2. Operations on `maxbig()` is what is causing HiPE to take a long time to compile. I verified the same issue happens today in `big_SUITE`, which is from where `maxbig()` implementation comes from. Here is a minimum file that reproduces the issue if someone from HiPE is interested in investigating it:

    ```erlang
    -module(foo).
    -export([run/0]).
    -compile(native).

    run() ->
        Maxbig = maxbig(),
        {'EXIT', {system_limit, [{erlang, '+', [Maxbig, 1], _} | _]}} = (catch Maxbig + 1),
        ok.

    maxbig() ->
        %% We assume that the maximum arity is (1 bsl 19) - 1.
        Ws = erlang:system_info(wordsize),
        (((1 bsl ((16777184 * (Ws div 4))-1)) - 1) bsl 1) + 1.
    ```

garazdawi commented:
we suspected that HiPE already had differences in how exceptions are handled, but it is good to have it confirmed. Leave it for now and I'll get back to you with the results from the test runs.

garazdawi commented:
There are three testcases in compiler that fail due to the changes in this pr.

trycatch_SUITE:eclectic, trycatch_no_opt_SUITE:eclectic, trycatch_post_opt_SUITE:eclectic

josevalim commented:
@garazdawi I have fixed and force pushed, thanks! They were failing simply because the test was not accounting for the new entries.
